stl/inc/ranges

```diff
@@ -728,6 +738,432 @@ namespace ranges {
         using all_t = decltype(all(_STD declval<_Rng>()));
     } // namespace views
 
+    // CLASS TEMPLATE ranges::take_view
+    template <view _Vw>
+    class take_view : public view_interface<take_view<_Vw>> {
+    private:
+        /* [[no_unique_address]] */ _Vw _Range{};
+        range_difference_t<_Vw> _Count = 0;
+
+        template <bool _Const, bool _Wrapped = true>
+        class _Sentinel {
+        private:
+            template <bool, bool>
+            friend class _Sentinel;
+
+            using _Base_Ty       = _Maybe_const<_Const, _Vw>;
+            using _Base_sentinel = _Maybe_wrapped<_Wrapped, sentinel_t<_Base_Ty>>;
+            using _Counted_Iter  = counted_iterator<_Maybe_wrapped<_Wrapped, iterator_t<_Base_Ty>>>;
+
+            _Base_sentinel _Last{};
+
+        public:
+            _Sentinel() = default;
+
+            constexpr explicit _Sentinel(_Base_sentinel _Last_) noexcept(
+                is_nothrow_move_constructible_v<_Base_sentinel>) // strengthened
+                : _Last(_STD move(_Last_)) {}
+
+            // clang-format off
+            constexpr _Sentinel(_Sentinel<!_Const, _Wrapped> _That)
+                noexcept(is_nothrow_constructible_v<_Base_sentinel, _Maybe_wrapped<_Wrapped, sentinel_t<_Vw>>>) // strengthened
+                requires _Const && convertible_to<_Maybe_wrapped<_Wrapped, sentinel_t<_Vw>>, _Base_sentinel>
+                : _Last(_STD move(_That._Last)) {}
+            // clang-format on
+
+            constexpr _Base_sentinel base() const
+                noexcept(is_nothrow_copy_constructible_v<_Base_sentinel>) /* strengthened */ {
+                return _Last;
+            }
+
+            friend constexpr bool operator==(const _Counted_Iter& _Left, const _Sentinel& _Right) {
+                return _Left.count() == 0 || _Left.base() == _Right._Last;

```

Comment:
So this  machinery is to prevent that we increment beyond the end of the range. 

However, this is already done through `counted_iterator::oeprato++`. The difference is that a non sized range in release will return early when we reach the end of the range, wheras a sized range will happily keep incrementing.

Note that `counted_iterator::operator==(default_sentinel_t)` only checks `count() == 0`

Is this really what we want?

Reply:
`take_view::begin()` takes the size of range into account when the range is sized. That is, for a sized range `r`, the `counted_iterator` returned from `take_view(r, n).begin()` always has a `count()` that's less or equal to `size(r)`.

```
auto r1 = std::list{1, 2, 3} | take(42); // r1.begin().count() is 3
auto r2 = null_terminated{"asd"} | take(42); // r2.begin().count() is 42
```

Reply:
Remember when we were discussing the `counted_iterator` constructor and I said "we don't want to require that `[i, i+ n)` is a valid range because it's not guaranteed that the entire range will be traversed?" This is exactly such a use case.

There are three cases that `take_view` handles differently in `begin` and `end`
1. Underlying range is sized and random-access: `begin` returns the underlying range's first iterator, and `end` returns the same value plus the minimum of the underlying range's size and the `take_view`'s count. The `take_view`s iterator and sentinel types are the same as the underlying range's iterator type, so `take_view::_Sentinel` is unused.
2. Underlying range is sized (but not random-access): We can again calculate the actual size of the `take_view` as the minimum of its count and the size of the underlying range, but we can't easily compute an end iterator as above. So, `begin` returns a `counted_iterator` initialized with the underlying range's first iterator and the calculated size, and `end` returns a `default_sentinel`. `take_view::_Sentinel` is again unused.
3. Underlying range is not sized, so we can't calculate the size of the `take_view` ahead of time. If `i` is the first iterator value of the underlying range, we want to stop when `++i` equals the underlying range's sentinel _or_ when we've incremented `i` "count" times. We're effectively iterating over two ranges at once: `[first, last)` (the underlying range of unknown length) and `[counted_iterator(first, count), default_sentinel)` (the counted range `first + [0, count)`). We encode those two ranges in a `counted_iterator` and a custom sentinel whose equality checks whether the `counted_iterator` has reached the end of either of those two ranges.

 For a non-sized underlying range `take_view` doesn't know the actual size of the range adapted, it only knows it wants to limit that size to some threshold. For this case we e

Reply:
Yeah, I just completely missread that the size() member takes the minimum of the two